src/mono/wasm/runtime/js-to-cs.ts

```diff
@@ -244,30 +246,22 @@ export function _wrap_js_thenable_as_task_root(thenable: Promise<any>, resultRoo
     // With more complexity we could recover original instance when this Task is marshaled back to JS.
     // TODO optimization: return the tcs.Task on this same call instead of _get_tcs_task
     const tcs_gc_handle = corebindings._create_tcs();
+    const holder: any = { tcs_gc_handle };
+    setup_managed_proxy(holder, tcs_gc_handle);
     thenable.then((result) => {
         corebindings._set_tcs_result_ref(tcs_gc_handle, result);
         // let go of the thenable reference
         mono_wasm_release_cs_owned_object(thenable_js_handle);
 
-        // when FinalizationRegistry is not supported by this browser, we will do immediate cleanup after promise resolve/reject
-        if (!_use_finalization_registry) {
-            corebindings._release_js_owned_object_by_gc_handle(tcs_gc_handle);
-        }
+        teardown_managed_proxy(holder, tcs_gc_handle); // this holds holder alive for finalizer, until the promise is freed
     }, (reason) => {
         corebindings._set_tcs_failure(tcs_gc_handle, reason ? reason.toString() : "");
         // let go of the thenable reference
         mono_wasm_release_cs_owned_object(thenable_js_handle);
 
-        // when FinalizationRegistry is not supported by this browser, we will do immediate cleanup after promise resolve/reject
-        if (!_use_finalization_registry) {
-            corebindings._release_js_owned_object_by_gc_handle(tcs_gc_handle);
-        }
+        teardown_managed_proxy(holder, tcs_gc_handle); // this holds holder alive for finalizer, until the promise is freed
```

Comment:
is it possible for _set_tcs_failure or _set_tcs_result to throw and stop us from reaching teardown and release? it seems possible and would probably cause bad behavior

Reply:
`_set_tcs_result_ref` does not propagate user's exception from whoever is awaiting the task.

It could only throw if runtime is in some terrible state already. In such case, I'm not sure what is the correct recovery.
As it is, the exception would bubble to main loop, as this is promise callback.

I at lest moved both `mono_wasm_release_cs_owned_object` and `teardown_managed_proxy` to `finally()`